src/Passes/Source/RuleTransformationPass/Profile.cpp

```diff
@@ -442,14 +440,32 @@ namespace quantum
             function->clearGC();
             function->clearMetadata();
 
+            // Deleteing blocks in reverse order
+            std::vector<llvm::BasicBlock*> blocks;
             for (auto& block : *function)
             {
-                llvm::errs() << "REMOVING" << block << "\n";
-                // Removing all instructions
+                blocks.push_back(&block);
+            }
+
+            // Removing backwards to avoid segfault
+            for (auto it = blocks.rbegin(); it != blocks.rend(); ++it)
+            {
+                auto& block = **it;
+                // Deleting instructions in reverse order (needed bauase it is a DAG structure)
+                std::vector<llvm::Instruction*> instructions;
                 for (auto& instr : block)
                 {
+                    instructions.push_back(&instr);
+                }
+
+                // Removing all instructions
+                for (auto it2 = instructions.rbegin(); it2 != instructions.rend(); ++it2)
+                {
+                    auto& instr = **it2;
+                    llvm::errs() << " -- " << instr << "\n";
                     instr.replaceAllUsesWith(llvm::UndefValue::get(instr.getType()));
-                    to_delete.push_back(&instr);
+                    instr.eraseFromParent();
+                    // to_delete.push_back(&instr);
                 }
 
                 // Removing all block references
```

Comment:
The line wasn't near enough to the ones that have been changed for me to comment on it directly, but it looks like the `replaceAllUsesWith` function for the blocks gets called twice. Since it is called on line 472, shouldn't it not need to be called on line 482? It seems like the block removal could be rearranged a bit to match the same pattern as the new instruction removal pattern you introduce here, but it's possible I'm missing a subtlety as to why it can't. If it can be refactored to match, then the `blocks_to_delete_` variable can be eliminated and the block erased from parent as soon as all uses are replaced.

Reply:
The one called on 472 is for the block whereas here we are replacing the instruction. I am not sure that I follow the suggestion - may it was a result of the original PR looking cluttered due to unmerged dependencies? If not, could you elaborate a bit?